### PR TITLE
Fix/ certificate download Bug

### LINKED ISSUE #764: bug: Invalid Certificate being downloaded
currently If we attempt to download the Certificate before its generated, then it still downloads a png which is corrupt since the it was downloaded before generating

https://github.com/code100x/cms/assets/136328410/5f9ca8e6-b5cf-4f18-b927-a36c4cb6e685

1. go to /certificate endpoint 
2. download the certificate before it appears
(enable throttle to fast/slow 3g to reproduce if you have very good internet)

**Expected behavior**
Waits for the certificate to be generated and then downloads it

 - Browser: chrome
 - Version 125.0.6422.113 (Official Build) (arm64)

### PR DESCRIPTION
**Pull request by nischal-shetty2**
### PR Fixes:
- 1 Waits for Certificate to first generate and only then downloads the Certificate
- 2 displays Generating while certificate is being generated

https://github.com/code100x/cms/assets/136328410/fa0f96b8-eb5d-4763-b1c6-986c0f8b7d92

Resolves #764

### Checklist before requesting a review
- [x] I assure there is no similar/duplicate pull request regarding same issue
- [x] I have performed a self-review of my code

### CODE REVIEW

**Comment by nischal-shetty2** [comment]
@siinghd @hkirat could you review this please

File: src/components/Certificate.tsx [hunk 1]
```diff
@@ -34,77 +40,80 @@ export const CertificateComponent = ({
...
-  // const handleDownloadPDF = async () => {
-  //   const downloadUrl = certificatePdfUrl;
-  //   const a = document.createElement('a');
-  //   a.href = downloadUrl!;
-  //   a.download = 'certificate.pdf';
-  //   document.body.appendChild(a);
-  //   a.click();
-  //   document.body.removeChild(a);
-  // };
+  useEffect(() => {
```
Comment (siinghd) [inline src/components/Certificate.tsx:43]: Why this?

File: src/components/Certificate.tsx [hunk 2]
```diff
@@ -11,14 +11,20 @@ import { Button } from './ui/button';
...
+
+type ActionType = 'downloadPNG' | 'shareLinkedIn' | 'shareTwitter';
 
 export const CertificateComponent = ({
   certificateId,
   certificateSlug,
   course,
   userName,
 }: OneCertificate & { userName: string }) => {
+  const [triggeredAction, setTriggeredAction] = useState<ActionType | null>(
```
Comment (siinghd) [inline src/components/Certificate.tsx:24]: What is this?

**Review by siinghd — CHANGES_REQUESTED** [review]
I think we dont need any of this, please just use state 

const [loading,setLoading] = useState(false);

in the functions just set it true initially and false at the end

**Comment by nischal-shetty2** [comment]
> I think we dont need any of this, please just use state
> 
> const [loading,setLoading] = useState(false);
> 
> in the functions just set it true initially and false at the end

had made a pr with this exact solution but thought it would be better if the user can click download and only after the certificate is generated does the download happen. ill make the requested changes and update @siinghd

**Comment by nischal-shetty2** [comment]
@siinghd @hkirat updated code with requested changes

https://github.com/code100x/cms/assets/136328410/68bf4d6c-e17a-4679-8f80-fcf4d12eeef4

File: src/components/Certificate.tsx (hunk 1 above)
Comment (nischal-shetty2) [reply src/components/Certificate.tsx:39]: changed in latest commit

File: src/components/Certificate.tsx [hunk 3]
```diff
@@ -34,15 +36,11 @@ export const CertificateComponent = ({
...
-  //   document.body.removeChild(a);
-  // };
+  useEffect(() => {
+    if (certificateImageUrl) {
+      setLoading(false);
+    }
+  }, [certificateImageUrl]);
 
   const handleDownloadPNG = async () => {
     const downloadUrl = certificateImageUrl;
```
Comment (siinghd) [inline src/components/Certificate.tsx:46]: we can also here add setLoading(true)

and at the end of the function set loading to false?

this way if the user clicks download multiple times the image doesnt get dowaloaded multiple times right ? 

do it for all the three functions

Comment (nischal-shetty2) [reply src/components/Certificate.tsx:46]: but then the buttons for share on linkedin and twitter would be disabled as well right? did you mean for me to make three states for three functions seperately?

Comment (nischal-shetty2) [reply src/components/Certificate.tsx:46]: @siinghd should i put a 1 second timeout after the click to fix this instead? or go with the 3 separate states? or did you mean some other way?

Comment (siinghd) [reply src/components/Certificate.tsx:46]: > @siinghd should i put a 1 second timeout after the click to fix this instead? or go with the 3 separate states? or did you mean some other way?

You already have loading state

Comment (nischal-shetty2) [reply src/components/Certificate.tsx:46]: > You already have loading state

@hkirat yes but that will disable all three of the buttons after clicking on download

Comment (nischal-shetty2) [reply src/components/Certificate.tsx:46]: @siinghd

File: src/components/Certificate.tsx [hunk 4]
```diff
@@ -53,58 +53,71 @@ export const CertificateComponent = ({
...
+    setLoading(true);
     const tweetText = `I just earned the "${course.title}" certificate on 100xDevs! Check it out:`;
     const certificateUrl = `${window.location.origin}/certificate/verify/${certificateSlug}`;
     const shareUrl = `https://twitter.com/intent/tweet?text=${encodeURIComponent(tweetText)}&url=${encodeURIComponent(certificateUrl)}`;
     window.open(shareUrl);
+    setLoading(false);
   };
 
-  console.log('Url is : ', certificatePdfUrl);
+  if (certificatePdfUrl) {
```
Comment (siinghd) [inline src/components/Certificate.tsx:81]: remove this please

**Review by siinghd — CHANGES_REQUESTED** [review]

File: src/components/Certificate.tsx (hunk 4 above)
Comment (nischal-shetty2) [reply src/components/Certificate.tsx:81]: > remove this please

sure, you meant the if check right? or the whole console.log?? @siinghd

File: src/components/Certificate.tsx [hunk 5]
```diff
@@ -78,9 +78,7 @@ export const CertificateComponent = ({
     setLoading(false);
   };
 
-  if (certificatePdfUrl) {
-    console.log('Url is : ', certificatePdfUrl);
-  }
+  console.log('Url is : ', certificatePdfUrl);
```
Comment (siinghd) [inline src/components/Certificate.tsx:81]: sir remove the console please

**Review by siinghd — CHANGES_REQUESTED** [review]

**Comment by nischal-shetty2** [comment]
@siinghd done sir! got confused because of the line you wrote the review in which was at the if check and since the console already existed i thought it was necessary

**Comment by nischal-shetty2** [comment]
@siinghd done with all the changes you can merge this now

**Comment by nischal-shetty2** [comment]
@siinghd waiting to be merged

**Comment by nischal-shetty2** [comment]
@siinghd check it out , done with all the changes

**Pull request merged**